osu.Server.Spectator/Hubs/CountdownImplementation.cs

```diff
@@ -0,0 +1,119 @@
+    /// </summary>
+    public class CountdownImplementation
+    {
+        private readonly long roomId;
+        private readonly IMultiplayerServerMatchCallbacks hub;
+
+        private CancellationTokenSource? countdownStopSource;
+        private CancellationTokenSource? countdownFinishSource;
+        private Task countdownTask = Task.CompletedTask;
+
+        public CountdownImplementation(long roomId, IMultiplayerServerMatchCallbacks hub)
+        {
+            this.roomId = roomId;
+            this.hub = hub;
+        }
+
+        /// <summary>
+        /// Starts a new countdown, stopping any existing one.
+        /// </summary>
+        /// <param name="countdown">The countdown to start. The <see cref="MultiplayerRoom"/> will receive this object for the duration of the countdown.</param>
+        /// <param name="onComplete">A callback to be invoked when the countdown completes.</param>
+        public void Start(MultiplayerCountdown countdown, Func<ServerMultiplayerRoom, Task> onComplete)
+        {
+            Stop();
+
+            var stopSource = countdownStopSource = new CancellationTokenSource();
+            var finishSource = countdownFinishSource = new CancellationTokenSource();
+            var cancellationSource = CancellationTokenSource.CreateLinkedTokenSource(stopSource.Token, finishSource.Token);
+
+            Task lastCountdownTask = countdownTask;
+            countdownTask = start();
+
+            // Some sections in the following code require single-threaded execution mode. This is achieved by re-retrieving the room from the hub, forcing an exclusive lock on it.
+            async Task start()
+            {
+                // 1. Wait for the last countdown.
+                try
+                {
+                    await lastCountdownTask;
+                }
+                catch
+                {
+                    // Any failures in the last countdown should not prevent future countdowns from running.
+                }
+
+                // 2. Notify users that a new countdown has started.
+                using (var roomUsage = await hub.GetRoom(roomId))
+                {
+                    if (roomUsage.Item == null)
+                        return;
+
+                    // The countdown could have been cancelled in a separate request before this task was able to run.
+                    if (stopSource.IsCancellationRequested)
+                        return;
+
+                    roomUsage.Item.Countdown = countdown;
+                    await hub.SendMatchEvent(roomUsage.Item, new CountdownChangedEvent { Countdown = countdown });
+                }
+
+                // 3. Run the countdown.
+                try
+                {
+                    await Task.Delay(countdown.EndTime - DateTimeOffset.Now, cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                }
```

Comment:
How does this work? I would presume that if there's a cancellation we don't want the code following this step to run.